src/groups/heisenberg.jl

```diff
@@ -0,0 +1,255 @@
+    ::RealNumbers,
+) where {n}
+    Xⁱ[1:n] .= _heisenberg_a_view(M, X)
+    Xⁱ[(n + 1):(2 * n)] .= _heisenberg_b_view(M, X)
+    Xⁱ[2 * n + 1] = X[1, n + 2]
+    return Xⁱ
+end
+
+get_embedding(::HeisenbergGroup{n}) where {n} = Euclidean(n + 2, n + 2)
+
+function get_vector_orthonormal(::HeisenbergGroup{n}, p, Xⁱ, ::RealNumbers) where {n}
+    return [
+        0 Xⁱ[1:n] Xⁱ[2 * n + 1]
+        zeros(n, n + 1) Xⁱ[(n + 1):(2 * n)]'
+        zeros(1, n + 2)
+    ]
+end
+
+function get_vector_orthonormal!(::HeisenbergGroup{n}, X, p, Xⁱ, ::RealNumbers) where {n}
+    fill!(X, 0)
+    X[1, 2:(n + 1)] .= Xⁱ[1:n]
+    X[2:(n + 1), n + 2] .= Xⁱ[(n + 1):(2 * n)]
+    X[1, n + 2] = Xⁱ[2 * n + 1]
+    return X
+end
+
+function exp_lie!(M::HeisenbergGroup{n}, q, X) where {n}
+    copyto!(q, I)
+    a_view = _heisenberg_a_view(M, X)
+    b_view = _heisenberg_b_view(M, X)
+    q[1, 2:(n + 1)] .= a_view
+    q[2:(n + 1), n + 2] .= b_view
+    q[1, n + 2] = X[1, n + 2] + dot(a_view, b_view) / 2
+    return q
+end
+
+@doc raw"""
+    exp(M::HeisenbergGroup, p, X)
+
+Exponential map on the [`HeisenbergGroup`](@ref) `M` with the left-invariant metric.
+"""
+exp(M::HeisenbergGroup, p, X)
+
+function exp!(M::HeisenbergGroup{n}, q, p, X) where {n}
+    q .= p * exp_lie(M, X)
+    return q
+end
+
+@doc raw"""
+    injectivity_radius(M::HeisenbergGroup)
+
+Return the injectivity radius on the [`HeisenbergGroup`](@ref) `M`, which is ``∞``.
+"""
+injectivity_radius(::HeisenbergGroup) = Inf
+
+function inner(M::HeisenbergGroup{n}, p, X, Y) where {n}
+    X_a_view = _heisenberg_a_view(M, X)
+    X_b_view = _heisenberg_b_view(M, X)
+    Y_a_view = _heisenberg_a_view(M, Y)
+    Y_b_view = _heisenberg_b_view(M, Y)
+    return dot(X_a_view, Y_a_view) +
+           dot(X_b_view, Y_b_view) +
+           X[1, 2 * n + 1] * Y[1, 2 * n + 1]
+end
+
+@doc raw"""
+    log(G::HeisenbergGroup, p, q)
+
+Compute the logarithmic map on the [`HeisenbergGroup`](@ref) group.
+
+"""
+log(::HeisenbergGroup, p, q)
+
+function log!(G::HeisenbergGroup{n}, X, p, q) where {n}
+    X .= log_lie(G, p \ q)
+    return X
+end
+
+function log_lie!(M::HeisenbergGroup{n}, X, p) where {n}
+    fill!(X, 0)
+    view_a_X = _heisenberg_a_view(M, X)
+    view_b_X = _heisenberg_b_view(M, X)
+    view_a_X .= _heisenberg_a_view(M, p)
+    view_b_X .= _heisenberg_b_view(M, p)
+    X[1, n + 2] = p[1, n + 2] - dot(view_a_X, view_b_X) / 2
+    return X
+end
+
+function log_lie!(::HeisenbergGroup, X, ::Identity{MultiplicationOperation})
+    fill!(X, 0)
+    return X
+end
+
+manifold_dimension(::HeisenbergGroup{n}) where {n} = 2 * n + 1
+
+parallel_transport_to(::HeisenbergGroup, p, X, q) = X
+
+parallel_transport_to!(::HeisenbergGroup, Y, p, X, q) = copyto!(Y, X)
+
+function project(M::HeisenbergGroup{n}, p) where {n}
+    return [
+        1 p[1, 2:(n + 2)]'
+        zeros(n, 1) Matrix(I, n, n) _heisenberg_b_view(M, p)
+        zeros(1, n + 1) 1
+    ]
+end
+function project(M::HeisenbergGroup{n}, p, X) where {n}
```

Comment:
can we maybe add doc strings here? I assume this is just projection with the Euclidean metric from the embedding, i.e. setting all necessary things to zero/one?

Reply:
same for the project above :)